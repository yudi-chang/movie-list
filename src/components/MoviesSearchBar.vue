<template>
  <div class="p-8 search-bar-wrapper">
    <input v-model="titleSearch" @input="debouncedSearch">
    <img src="/images/search-icon.png" alt="Search" />
  </div>
</template>

<script setup lang="ts">
import { ref, inject } from 'vue';
import { useDebounce } from '@/composables/debounce'
import type { MoviesFetchPayload } from '@/types';

// inject
const fetchMovies = inject<(payload: MoviesFetchPayload) => void>('fetchMovies')

// reactive
const titleSearch = ref<string>('')

// methods
const fetchMoviesWithPayload = (): void => {
  if (fetchMovies) {
    const payload: MoviesFetchPayload = {
      Title: titleSearch.value,
      page: 1,
    }
    fetchMovies(payload)
  }
}

// use debounce composable
const { debounce, isWaiting } = useDebounce<string>(() => {
  fetchMoviesWithPayload()
}, 450)

const debouncedSearch = () => {
  debounce()
}
</script>

<style scoped lang="scss">
.search-bar-wrapper {
  background-color: $primary-color;
  position: relative;
  border-radius: 5px;

  input {
    width: 100%;
    height: 50px;
    border-radius: 5px;
    position: relative;
    z-index: 0;
    background-color: #212a31;
    color: $secondary-color;
    transition: border 0.2s ease-in-out;
    padding: 0 45px 0 15px;
  }

  img {
    position: absolute;
    top: 50%;
    right: 25px;
    transform: translateY(-50%);
    height: 20px;
    z-index: 999;
  }
}
</style>