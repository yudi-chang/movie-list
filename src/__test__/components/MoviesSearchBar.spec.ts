import { mount } from '@vue/test-utils';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import MoviesSearchBar from '@/components/MoviesSearchBar.vue';
import type { MoviesFetchPayload } from '@/types';

const mockFetchMovies = vi.fn();

describe('MoviesSearchBar', () => {
  const defaultParams =  {
    global: {
      provide: {
        fetchMovies: mockFetchMovies,
      },
    },
  }

  const createWrapper = (params = defaultParams) => {
    return mount(MoviesSearchBar, params);
  };

  beforeEach(() => {
    vi.useFakeTimers();
    mockFetchMovies.mockClear();
  });

  it('should renders correctly', () => {
    const wrapper = createWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should calls fetchMovies when input changes', async () => {
    const wrapper = createWrapper();
    const input = wrapper.find('[data-test="search-field"]');

    await input.setValue('water');

    vi.advanceTimersByTime(450);

    expect(mockFetchMovies).toHaveBeenCalledWith({
      Title: 'water',
      page: 1,
    } as MoviesFetchPayload);
  });

  it('should not call fetchMovies if fetchMovies is not provided', async () => {
    const wrapper = createWrapper({} as any);
    const input = wrapper.find('input');

    await input.setValue('water');
    vi.advanceTimersByTime(450);

    expect(mockFetchMovies).not.toHaveBeenCalled();
  });

  it('should debounces input to avoid excessive API calls', async () => {
    const wrapper = createWrapper();
    const input = wrapper.find('input');

    await input.setValue('w');
    await input.setValue('wat');
    await input.setValue('water');

    // before debounce triggers
    vi.advanceTimersByTime(449);
    expect(mockFetchMovies).not.toHaveBeenCalled();

    // debounce triggered
    vi.advanceTimersByTime(1);
    expect(mockFetchMovies).toHaveBeenCalledTimes(1);
    expect(mockFetchMovies).toHaveBeenCalledWith({
      Title: 'water',
      page: 1,
    } as MoviesFetchPayload);
  });
});
