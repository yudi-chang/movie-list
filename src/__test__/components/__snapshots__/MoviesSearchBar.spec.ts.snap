// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MoviesSearchBar > should renders correctly 1`] = `
<div
  class="p-8 search-bar-wrapper"
  data-v-26d1eaa1=""
>
  <input
    data-test="search-field"
    data-v-26d1eaa1=""
  />
  <img
    alt="Search"
    data-v-26d1eaa1=""
    src="/images/search-icon.png"
  />
</div>
`;
